components/_impl/workers/subprocess_worker.py

```diff
@@ -0,0 +1,346 @@
+            raise NotImplementedError("SubprocessWorker does not support `in_memory`")
+
+        # NB: we convert the bytes to a hex string to avoid encoding issues.
+        self._run(f"""
+            {name} = {subprocess_rpc.WORKER_IMPL_NAMESPACE}["marshal"].loads(
+                bytes.fromhex({repr(marshal.dumps(value).hex())})
+            )
+        """)
+
+    def load(self, name: str) -> typing.Any:
+        self._run(f"""
+            {subprocess_rpc.WORKER_IMPL_NAMESPACE}["load_pipe"].write(
+                {subprocess_rpc.WORKER_IMPL_NAMESPACE}["marshal"].dumps({name})
+            )
+        """)
+
+        return marshal.loads(self._load_pipe.read())
+
+    @property
+    def in_process(self) -> bool:
+        return False
+
+    @property
+    def alive(self) -> bool:
+        return self._alive and self._proc.poll() is None
+
+    def _bootstrap_worker(self) -> None:
+        """Import subprocess_rpc in the worker, and start the work loop.
+
+        Commands are executed by writing to `self._input_pipe`, and waiting for
+        a response on `self._output_pipe`. This presumes, however, that there
+        is a worker doing the opposite: listening to the input pipe and writing
+        to the output pipe. At startup `self._proc` is a simple interactive
+        Python process, so we have to bootstrap it to start the work loop or
+        else `self._run` will hang waiting for jobs to be processed.
+        """
+
+        # NB: This gets sent directly to `self._proc`'s stdin, so it MUST be
+        #     a single expression and may NOT contain any empty lines. (Due to
+        #     how Python processes commands.)
+        bootstrap_command = textwrap.dedent(f"""
+            try:
+                import marshal
+                import sys
+                sys_path_old = list(sys.path)
+                sys.path = marshal.loads(
+                    bytes.fromhex({repr(marshal.dumps(sys.path).hex())})
+                )
+                # The parent gets priority, but a subclass could set PYTHONPATH
+                # so we have to respect extra paths.
+                sys.path.extend([i for i in sys_path_old if i and i not in sys.path])
+                from components._impl.workers import subprocess_rpc
+                output_pipe = subprocess_rpc.Pipe(
+                    write_handle={self._output_pipe.write_handle})
+                output_pipe.write(subprocess_rpc.BOOTSTRAP_IMPORT_SUCCESS)
+                subprocess_rpc.run_loop(
+                    input_handle={self._input_pipe.read_handle},
+                    output_pipe=output_pipe,
+                    load_handle={self._load_pipe.write_handle},
+                )
+            except:
+                sys.exit(1)
+        """).strip()
+
+        if self._proc.poll() is not None:
+            raise ValueError("Process has already exited.")
+
+        proc_stdin = self._proc.stdin
+        assert proc_stdin is not None
+
+        self._log_cmd(bootstrap_command)
```

Comment:
I'm trying to understand why _log_cmd -> read doesn't go through a common shared functionality with _run?

Reply:
The hang up is that `_run` requires that the work loop is already running on the worker, so we can't use it in the bootstrapping phase.